### PR TITLE
HEVC 10bit encoder and P010/I010 in put enable

### PR DESCRIPTION
**Pull request by QuPengfei**
this enable the P010/I010 format support and HEVC 10bit encoder

### CODE REVIEW

File: vaapi/vaapisurfaceallocator.cpp [hunk 1]
```diff
@@ -33,24 +33,35 @@ VaapiSurfaceAllocator::VaapiSurfaceAllocator(VADisplay display, uint32_t extraSi
 
 YamiStatus VaapiSurfaceAllocator::doAlloc(SurfaceAllocParams* params)
 {
+    uint32_t rtFormat = VA_RT_FORMAT_YUV420;
+    VASurfaceAttrib attrib;
```
Comment (xuguangxin) [inline vaapi/vaapisurfaceallocator.cpp:36]: Better memset any va structure before we use it, It maybe add fields in future.

Comment (xuguangxin) [inline vaapi/vaapisurfaceallocator.cpp:36]: Could we move this to line 58?

File: interface/VideoCommonDefs.h [hunk 2]
```diff
@@ -69,6 +69,8 @@ extern "C" {
          ((uint32_t)(uint8_t)(ch2) << 16)  | ((uint32_t)(uint8_t)(ch3) << 24))
 
 #define YAMI_FOURCC_NV12 YAMI_FOURCC('N','V','1','2')
+#define YAMI_FOURCC_P010 YAMI_FOURCC('P','0','1','0')
+#define YAMI_FOURCC_I010 YAMI_FOURCC('I','0','1','0')
```
Comment (xuguangxin) [inline interface/VideoCommonDefs.h]: Do we really need this in core library level. Could we support P010 only? And let yami-utils do I010 to P010 conversion?

File: encoder/vaapiencoder_hevc.cpp [hunk 3]
```diff
@@ -866,6 +867,7 @@ VaapiEncoderHEVC::VaapiEncoderHEVC():
 {
     m_videoParamCommon.profile = VAProfileHEVCMain;
     m_videoParamCommon.level = 51;
+    m_videoParamCommon.bitDepth = 8;
```
Comment (xuguangxin) [inline encoder/vaapiencoder_hevc.cpp]: better put this to base class. Every codec need this. Actually 

> m_pool = SurfacePool::create(m_alloc, ((m_videoParamCommon.bitDepth == 8)?YAMI_FOURCC_NV12:YAMI_FOURCC_P010), (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);

will wrong if we do not set this in base

File: encoder/vaapiencoder_base.cpp [hunk 4]
```diff
@@ -498,7 +661,7 @@ bool VaapiEncoderBase::initVA()
 
     int32_t surfaceWidth = ALIGN16(m_videoParamCommon.resolution.width);
     int32_t surfaceHeight = ALIGN16(m_videoParamCommon.resolution.height);
-    m_pool = SurfacePool::create(m_alloc, YAMI_FOURCC_NV12, (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);
+    m_pool = SurfacePool::create(m_alloc, ((m_videoParamCommon.bitDepth == 8)?YAMI_FOURCC_NV12:YAMI_FOURCC_P010), (uint32_t)surfaceWidth, (uint32_t)surfaceHeight, m_maxOutputBuffer);
```
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp]: Could we return false for other bitDepth?

File: common/utils.cpp [hunk 5]
```diff
@@ -130,13 +131,24 @@ bool getPlaneResolution(uint32_t fourcc, uint32_t pixelWidth, uint32_t pixelHeig
...
                 width[1]  = w + (w & 1);
                 height[1] = (h + 1) >> 1;
                 planes = 2;
+            } else if (fourcc == VA_FOURCC_P010) {
+                width[0] = w * 2;
+                width[1]  = w * 2;
+                height[1] = (h + 1) >> 1;
+                planes = 2;
+            } else if (fourcc == VA_FOURCC('I', '0', '1', '0')) {
+                width[1] = width[2] = (w + 1) >> 1;
```
Comment (xuguangxin) [inline common/utils.cpp]: little confused, what is the diffrence between I420 and I010, why we can't merge this with else branch.Seems they are same code

File: encoder/vaapiencoder_base.cpp [hunk 6]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+
+    default:
+        bpp = 1;
+        break;
+    }
+
+    return bpp;
+}
+
+static bool loadFrameToSurface(VAImage surface_image,uint8_t* dst,uint8_t* src,uint32_t fourcc)
```
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp:376]: why we need this? could handle it with getPlaneResolution and copyImage?

**Comment by xuguangxin** [comment]
Hi @QuPengfei , thanks for your patch. We should suggest other member to implement new features like this; it will save efforts for all of us. ;).

**Comment by lizhong1008** [comment]
Seems need to change these commit names.

File: common/utils.cpp (hunk 5 above)
Comment (lizhong1008) [inline common/utils.cpp]: For 10bits pixel, it should need 16bits with the lowest 6 bits set to zero. so,width[0]= 2*w,  width[1] = width[2] = w, right?

File: encoder/vaapiencoder_base.cpp [hunk 7]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+{
+    uint32_t bpp;
+
+    switch (fourcc) {
+    case VA_FOURCC_NV12:
+    case VA_FOURCC_YV12:
+    case VA_FOURCC_I420:
+        bpp = 1;
+        break;
+
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp:362]: line 356 to 360 can be removed.

File: encoder/vaapiencoder_base.cpp [hunk 8]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+        if ((fourcc == VA_FOURCC('I', '0', '1', '0')) ||
+            (fourcc == VA_FOURCC_I420)) {
+            u_src = src + surface_image.width * surface_image.height * bpp;
+            v_src = src + surface_image.width * surface_image.height * 5 / 4 * bpp;
+        } else {
+            ERROR("only support I010/I420 ");
+            return false;
+        }
+
+        y_dst = (unsigned char *)((unsigned char*)surface_p + surface_image.offsets[0]);
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp]: would better use uint8\* instead of unsigned char *. And the first (unsigned char *) is not necessary.

File: encoder/vaapiencoder_base.cpp [hunk 9]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+        uint32_t bpp = get_bpp_in_1stplane_by_fourcc(surface_image.format.fourcc);
+
+        y_src = src;
+        if ((fourcc == VA_FOURCC('I', '0', '1', '0')) ||
+            (fourcc == VA_FOURCC_I420)) {
+            u_src = src + surface_image.width * surface_image.height * bpp;
+            v_src = src + surface_image.width * surface_image.height * 5 / 4 * bpp;
+        } else {
+            ERROR("only support I010/I420 ");
+            return false;
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp]: We can return false if fourcc is not I010 or I420 at the first of this function to make code more clear.

File: encoder/vaapiencoder_base.cpp [hunk 10]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+                 surface_image.format.fourcc == VA_FOURCC_P016) {
+            u_dst = (unsigned char *)((unsigned char*)surface_p + surface_image.offsets[1]);
+            v_dst = u_dst;
+        }
+        else
+        {
+            ERROR("don't support this surface fcc");
+            return false;
+        }
+
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp:420]: No need these code.

File: encoder/vaapiencoder_base.cpp [hunk 11]
```diff
@@ -341,6 +348,146 @@ static bool copyImage(uint8_t* destBase,
...
+        /* Y plane, directly copy */
+        for (row = 0; row < surface_image.height; row++) {
+            if(bpp == 2){
```
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp:423]: would better use forcc ==10bits,  instead of bpp==2. If one day we need to support 16 bits pixel(here VA_FOURCC_P016 is added), the bpp still equal to 2, but the logic is different

File: encoder/vaapiencoder_base.cpp (hunk 4 above)
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp]: why forcc is P010 if is not 8 bits? it can be VA_FOURCC_P016.

File: encoder/vaapiencoder_hevc.cpp (hunk 3 above)
Comment (lizhong1008) [inline encoder/vaapiencoder_hevc.cpp]: Agree with @xuguangxin

File: common/utils.cpp (hunk 5 above)
Comment (xuguangxin) [inline common/utils.cpp]: I mean seems L149 to L152 is duplicate for L153 to L155, why we can't merge them to else branch?

File: encoder/vaapiencoder_base.cpp (hunk 11 above)
Comment (xuguangxin) [inline encoder/vaapiencoder_base.cpp:423]: Not sure why we need loadFrameToSurface. Could we just use getPlaneResolution and copyImage like other formats?

File: common/utils.cpp (hunk 5 above)
Comment (lizhong1008) [inline common/utils.cpp]: L149 to L 151 is error, I010 is differenct from I420.

**Comment by lizhong1008** [comment]
Does this patch series support hevc 10 bits encoding? seems bitDepth is always 8, and profile is always HEVC main profile, not Main 10 profile.

File: encoder/vaapiencoder_base.cpp [hunk 12]
```diff
@@ -274,16 +274,23 @@ SurfacePtr VaapiEncoderBase::createNewSurface(uint32_t fourcc)
     uint32_t rtFormat;
     SurfacePtr surface;
 
+
```
Comment (chivakker) [inline encoder/vaapiencoder_base.cpp]: unnecessary change

File: vaapi/vaapisurfaceallocator.cpp [hunk 13]
```diff
@@ -33,24 +33,35 @@ VaapiSurfaceAllocator::VaapiSurfaceAllocator(VADisplay display, uint32_t extraSi
...
         return YAMI_INVALID_PARAM;
     uint32_t size = params->size;
     uint32_t width = params->width;
     uint32_t height = params->height;
     if (!width || !height || !size)
         return YAMI_INVALID_PARAM;
-    if (params->fourcc != YAMI_FOURCC_NV12) {
-        ERROR("fix me for 10 bits");
+    if (params->fourcc != YAMI_FOURCC_NV12 && params->fourcc != YAMI_FOURCC_P010) {
+        ERROR("only support NV12 and P010");
```
Comment (chivakker) [inline vaapi/vaapisurfaceallocator.cpp:53]: Needs clang format

**Comment by QuPengfei** [comment]
thanks for your comments. guanxing,lizhong and Daniel .

File: vaapi/vaapisurfaceallocator.cpp (hunk 1 above)
Comment (QuPengfei) [inline vaapi/vaapisurfaceallocator.cpp:36]: agree to memset.  it is ok to move to line 58.

File: interface/VideoCommonDefs.h (hunk 2 above)
Comment (QuPengfei) [inline interface/VideoCommonDefs.h]: it is OK to only use P010. For I010, it is OK to use by YAMI_FOURCC('I','0','1','0')

File: encoder/vaapiencoder_hevc.cpp (hunk 3 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_hevc.cpp]: agree

File: encoder/vaapiencoder_base.cpp (hunk 4 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp]: feed the VA_FOURCC_P016 when 10bit
use VA_FOURCC_NV12 as the default mode when bitDepth !=8

and if we add 12bit or more ,just add here in the later

File: common/utils.cpp (hunk 5 above)
Comment (QuPengfei) [inline common/utils.cpp]: zhong is correct. wrong width calculation for I010 here. 

```
        } else if (fourcc == VA_FOURCC('I', '0', '1', '0')) {
            width[0] = 2* w; 
            width[1] = width[2] = w;
            height[1] = height[2] = (h + 1) >> 1;
            planes = 3;
        } else {
```

File: encoder/vaapiencoder_base.cpp (hunk 6 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp:376]: it convert I010->P010 and I420->NV12. copyImage doesnot.

the driver can convert I420->NV12 when input is I420.  and it does not support I010->I420.

File: encoder/vaapiencoder_base.cpp (hunk 8 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp]: agree

File: encoder/vaapiencoder_base.cpp (hunk 7 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp:362]: it is better keep the codes there. it help.

File: vaapi/vaapisurfaceallocator.cpp (hunk 13 above)
Comment (QuPengfei) [inline vaapi/vaapisurfaceallocator.cpp:53]: agree

File: encoder/vaapiencoder_base.cpp (hunk 12 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp]: agree

File: encoder/vaapiencoder_base.cpp (hunk 11 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp:423]: @lizhong1008, from my understanding, the 16 bit  has the same logic as 10bits. i am little confused what is the difference here?

File: common/utils.cpp (hunk 5 above)
Comment (QuPengfei) [inline common/utils.cpp]: there is error when NV12
            if (fourcc == VA_FOURCC_NV12) {
                 width[1]  = w + (w & 1);
                 height[1] = (h + 1) >> 1;
                 planes = 2;
here width[1] = w;

File: encoder/vaapiencoder_base.cpp (hunk 11 above)
Comment (lizhong1008) [inline encoder/vaapiencoder_base.cpp:423]: @QuPengfei Dose 16 bits need "value <<= 6" in line 426?

File: encoder/vaapiencoder_base.cpp (hunk 9 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp]: agree

File: encoder/vaapiencoder_base.cpp (hunk 11 above)
Comment (QuPengfei) [inline encoder/vaapiencoder_base.cpp:423]: yes, 16bits does not need this line. here is a bug.

**Comment by lizhong1008** [comment]
@QuPengfei ,  any update for it?

**Comment by QuPengfei** [comment]
According to code reivew, the first version changes are reday on https://github.com/QuPengfei/libyami.

i also talk with guangxin last week. he hope to upstream JPEG related changes firstly. and then the patch still need modification based on it.

what is your plan for it?

**Comment by lizhong1008** [comment]
@QuPengfei , "the  JPEG related changes (https://github.com/01org/libyami/pull/537)" have been merged. You can rebased on it.

**Comment by lizhong1008** [comment]
@QuPengfei , any update for this pull request?

**Comment by xuguangxin** [comment]
this will rebase after my hevc decoder branch merged. Since many things have been changed.

**Comment by xuguangxin** [comment]
Hi @QuPengfei 
After some code refact, add p010 is pretty simply compare to previous patch. It only need one line code https://github.com/01org/libyami/pull/638/commits/788be3ba30e68c9332314be5165f239720e63f63. 
I have add to my hevc 10 bit decoder,Please rebase this on my patch set.
thanks

**Comment by xuguangxin** [comment]
close this since we have #668
